core/embed/trezorhal/common.c

```diff
@@ -25,13 +25,23 @@
 #include "display.h"
 #include "flash.h"
 #include "rand.h"
+#include "supervise.h"
 
 #include "stm32f4xx_ll_utils.h"
 
+#define COLOR_FATAL_ERROR RGB16(0x7F, 0x00, 0x00)
+
 // from util.s
-extern void shutdown(void);
+extern void shutdown_privileged(void);
 
-#define COLOR_FATAL_ERROR RGB16(0x7F, 0x00, 0x00)
+void shutdown(void) {
+#ifdef USE_SVC_SHUTDOWN
+  svc_shutdown();
+#else
+  // It won't work properly unless called from the privileged mode
+  shutdown_privileged();
+#endif
+}
```

Comment:
Maybe one suggestion to consider using define instead of a function here.

But maybe that's over-optimizing stuff and compiler does this anyway ...

```suggestion
#ifdef USE_SVC_SHUTDOWN
#define shutdown svc_shutdown
#else
// It won't work properly unless called from the privileged mode
#define shutdown shutdown_privileged
#endif
```

Reply:
> But maybe that's over-optimizing stuff and compiler does this anyway ...

Yes, compiler inlines it. I'd leave it as it is.